File: frontend/core/noteSequence.ts
import { SongChords, type ChordsSection, type Song } from "@/client";

export interface Note {
  note: string;
  rest: string;
  index: number;
  bassNote?: Note;
}

export const NOTES_TO_INDEXES: { [K: string]: number } = {
  C: 0,
  "C#": 1,
  Db: 1,
  D: 2,
  "D#": 3,
  Eb: 3,
  E: 4,
  F: 5,
  "F#": 6,
  Gb: 6,
  G: 7,
  "G#": 8,
  Ab: 8,
  A: 9,
  "A#": 10,
  Bb: 10,
  B: 11,
};

export const NORMALIZED_NOTES = Object.values(SongChords.key);
const ACCIDENALS = ["#", "b"];

export enum NoteSequenceErrMsg {
  NO_VALUE = "Ожидалось значение", // "Value expected",
  MULTIPLE_ACCIDENTALS = "Допустим только один знак альтерации", // "Only one accidental allowed",
  MULTPLE_BASS_NOTES = "Допустипа только одна басс-нота", // "Only one bass note allowed",
  INVALID_NOTE = "Несуществующая нота", // "Invalid note",
  SHOULD_START_WITH = "Строка должна начинаться с |", // "Line should start with |",
  SHOULD_END_WITH = "Строка должна заканчиваться на |", // "Line should end with |",
}

export class NoteSequenceValidationError extends Error {
  message: NoteSequenceErrMsg;
  value: string;

  constructor(message: NoteSequenceErrMsg, value: string) {
    super();
    this.message = message;
    this.value = value;
  }
}

function assert(expr: unknown, message: NoteSequenceErrMsg, value: string) {
  if (!expr) throw new NoteSequenceValidationError(message, value);
}

export function parseNote(value: string): Note {
  const val = value.trim().toLowerCase();
  assert(val, NoteSequenceErrMsg.NO_VALUE, value);

  const pureNote = val[0].toUpperCase();
  let accidental;
  let rawRest;

  if (val.length > 1 && ACCIDENALS.includes(val[1])) {
    accidental = val[1];
    rawRest = val.slice(2);
  } else {
    accidental = "";
    rawRest = val.slice(1);
  }

  let buildingBassNote = false;
  let bassNoteVal = "";
  let rest = "";

  for (const symbol of rawRest) {
    if (buildingBassNote) {
      bassNoteVal += symbol;
    } else {
      assert(
        !ACCIDENALS.includes(symbol),
        NoteSequenceErrMsg.MULTIPLE_ACCIDENTALS,
        value
      );

      if (symbol == "/") {
        buildingBassNote = true;
      } else {
        rest += symbol;
      }
    }
  }

  assert(
    !bassNoteVal.includes("/"),
    NoteSequenceErrMsg.MULTPLE_BASS_NOTES,
    value
  );
  const bassNote = bassNoteVal ? parseNote(bassNoteVal) : undefined;

  const note = pureNote + accidental;
  const index = NOTES_TO_INDEXES[note];
  assert(typeof index == "number", NoteSequenceErrMsg.INVALID_NOTE, value);

  return { note: note, rest: rest.trim(), index: index, bassNote: bassNote };
}

export function stringifyNote(note: Note): string {
  if (note.bassNote)
    return `${note.note}${note.rest}/${stringifyNote(note.bassNote)}`;
  return `${note.note}${note.rest}`;
}

export interface Bar {
  notes: Note[];
}

export function parseBar(value: string): Bar {
  const val = value.trim();
  assert(val, NoteSequenceErrMsg.NO_VALUE, value);
  return { notes: val.split(/\s+/).map(parseNote) };
}

export function stringifyBar(value: Bar): string {
  return value.notes.map(stringifyNote).join(" ");
}

export interface BarLine {
  bars: Bar[];
}

export function parseBarLine(value: string): BarLine | undefined {
  const val = value.trim();
  if (!val) return;
  assert(val.startsWith("|"), NoteSequenceErrMsg.SHOULD_START_WITH, value);
  assert(val.endsWith("|"), NoteSequenceErrMsg.SHOULD_END_WITH, value);
  const rawBars = val.slice(1, -1).trim().split("|");
  return { bars: rawBars.map(parseBar) };
}

export function stringifyBarLine(value: BarLine): string {
  const joinedBars = value.bars.map(stringifyBar).join(" | ");
  return `| ${joinedBars} |`;
}

type NoteSequence = BarLine[];

export function parseNoteSequence(value: string): NoteSequence {
  return value
    .split("\n")
    .map(parseBarLine)
    .filter((barLine): barLine is BarLine => !!barLine);
}

export function stringifyNoteSequence(noteSequence: NoteSequence): string {
  return noteSequence.map(stringifyBarLine).join("\n");
}

function mod(number: number, remainder: number): number {
  return ((number % remainder) + remainder) % remainder;
}

export function transposeNote(note: Note, halfSteps: number): Note {
  const index = mod(note.index + halfSteps, 12);
  const bassNote = note.bassNote
    ? transposeNote(note.bassNote, halfSteps)
    : undefined;

  return {
    note: NORMALIZED_NOTES[index],
    rest: note.rest,
    index: index,
    bassNote: bassNote,
  };
}

function transposeNoteSequence(
  noteSequence: NoteSequence,
  halfSteps: number
): NoteSequence {
  return noteSequence.map((barLine) => ({
    bars: barLine.bars.map((bar) => ({
      notes: bar.notes.map((note) => transposeNote(note, halfSteps)),
    })),
  }));
}

export function getHalfSteps(
  key: SongChords.key,
  targetKey: SongChords.key
): number {
  return NORMALIZED_NOTES.indexOf(targetKey) - NORMALIZED_NOTES.indexOf(key);
}

function transposeNotes(
  notes: string,
  key: SongChords.key,
  targetKey: SongChords.key
): string {
  return stringifyNoteSequence(
    transposeNoteSequence(
      parseNoteSequence(notes),
      getHalfSteps(key, targetKey)
    )
  );
}

export function transposeSong(
  song?: Song,
  targetKey?: SongChords.key
): ChordsSection[] | undefined {
  if (!song?.chords || !targetKey) return;
  const chords = song.chords;

  const transposeSection = (section: ChordsSection) => ({
    name: section.name,
    notes: transposeNotes(section.notes, chords.key, targetKey),
  });

  return chords.sections.map(transposeSection);
}

File: backend/core/note_sequence.py
from __future__ import annotations

from collections.abc import Iterable
from typing import Final, Optional

from pydantic import BaseModel, PydanticValueError
from typing_extensions import Self

NOTES_TO_INDEXES: Final = {
    "C": 0,
    "C#": 1,
    "Db": 1,
    "D": 2,
    "D#": 3,
    "Eb": 3,
    "E": 4,
    "F": 5,
    "F#": 6,
    "Gb": 6,
    "G": 7,
    "G#": 8,
    "Ab": 8,
    "A": 9,
    "A#": 10,
    "Bb": 10,
    "B": 11,
}

NORMALIZED_NOTES: Final = (
    "C",
    "C#",
    "D",
    "D#",
    "E",
    "F",
    "F#",
    "G",
    "G#",
    "A",
    "A#",
    "B",
)

ACCIDENTALS: Final = ("#", "b")


class Note(BaseModel):
    note: str
    rest: str
    index: int
    bass_note: Optional[Note] = None

    def __str__(self) -> str:
        if self.bass_note:
            return f"{self.note}{self.rest}/{str(self.bass_note)}"
        return f"{self.note}{self.rest}"

    @classmethod
    def parse(cls, value: str) -> Self:
        # "Bm" or " "
        val = value.strip().lower()
        assert val, ("Value expected", value)

        pure_note = val[0].upper()

        if len(val) > 1 and val[1] in ACCIDENTALS:  # has accidental
            accidental = val[1]
            raw_rest = val[2:]
        else:
            accidental = ""
            raw_rest = val[1:]

        building_bass_note = False
        bass_note_val = ""
        rest = ""

        for symbol in raw_rest:
            if building_bass_note:
                bass_note_val += symbol

            else:
                assert symbol not in ACCIDENTALS, ("Only one accidental allowed", value)

                if symbol == "/":  # bass note starts, for example, C# in A/C#
                    building_bass_note = True
                else:
                    rest += symbol

        assert "/" not in bass_note_val, ("Only one bass note allowed", value)
        bass_note = Note.parse(bass_note_val) if bass_note_val else None

        note = pure_note + accidental
        index = NOTES_TO_INDEXES.get(note)
        assert index is not None, ("Invalid note", value)

        return Note(note=note, rest=rest.strip(), index=index, bass_note=bass_note)


class Bar(BaseModel):
    notes: list[Note]

    def __str__(self) -> str:
        return " ".join(str(note) for note in self.notes)

    @classmethod
    def parse(cls, value: str) -> Self:
        # "Bm A" or " "
        val = value.strip()
        assert val, ("Value expected", value)
        return cls(notes=[Note.parse(note) for note in val.split()])


class BarLine(BaseModel):
    bars: list[Bar]

    def __str__(self) -> str:
        joined_bars = " | ".join(str(bar) for bar in self.bars)
        return f"| {joined_bars} |"

    @classmethod
    def parse(cls, value: str) -> Self | None:
        # "| G/C | Bm A |" or " "
        if not (val := value.strip()):
            return
        assert val.startswith("|"), ("Line should start with |", value)
        assert val.endswith("|"), ("Line should end with |", value)
        raw_bars = val[1:-1].strip().split("|")
        return cls(bars=[Bar.parse(bar) for bar in raw_bars])


class NoteSequenceValidationError(PydanticValueError):
    msg_template = '{message}: "{value}"'


def parse_note_sequence(value: str) -> list[BarLine]:
    def parse_bar_line(bar_line: str) -> BarLine | None:
        try:
            return BarLine.parse(bar_line)
        except AssertionError as exc:
            message, value = exc.args[0]
            assert isinstance(message, str)
            raise NoteSequenceValidationError(message=message, value=value)

    def parse_bar_lines(value: str) -> Iterable[BarLine]:
        for bar_line in value.splitlines():
            if result := parse_bar_line(bar_line):
                yield result

    return list(parse_bar_lines(value))


def stringify_note_sequence(bar_lines: list[BarLine]) -> str:
    return "\n".join(str(bar_line) for bar_line in bar_lines)
